File: Flask/modelo/Categoria.py
from modelo.Banco import Banco
from mysql.connector import Error

class Categoria:
    def __init__(self):
        self._id_categoria = None
        self._nome_categoria = None
        self._desconto_categoria = None
        self._id_setor = None

    def create(self):
        conexao = Banco.getConexao()
        if conexao:
            try:
                cursor = conexao.cursor()
                sql = "INSERT INTO categorias (nome_categoria, desconto_categoria, id_setor) VALUES (%s, %s, %s)"
                cursor.execute(sql, (self._nome_categoria, self._desconto_categoria, self._id_setor))
                conexao.commit()
                self._id_categoria = cursor.lastrowid
                return True
            except Error as e:
                print(f"Erro ao criar categoria: {e}")
                return False
            finally:
                cursor.close()

    def delete(self, id):
        conexao = Banco.getConexao()
        
        if conexao:
            cursor = None
            try:
                cursor = conexao.cursor()
                sql = "DELETE FROM categorias WHERE id_categoria = %s"
                cursor.execute(sql, (id,))
                conexao.commit()
                return cursor.rowcount > 0  # Retorna True se alguma linha foi afetada
            except Error as e:
                print(f"Erro ao deletar categoria: {e}")
                return False
            finally:
                if cursor:
                    cursor.close()

    def update(self):
        conexao = Banco.getConexao()
        
        if conexao:
            cursor = None
            try:
                cursor = conexao.cursor()
                sql = "UPDATE categorias SET nome_categoria = %s, desconto_categoria = %s, id_setor = %s WHERE id_categoria = %s"
                cursor.execute(sql, (self._nome_categoria, self._desconto_categoria, self._id_setor, self._id_categoria))
                conexao.commit()
                return cursor.rowcount > 0  # Retorna True se alguma linha foi afetada
            except Error as e:
                print(f"Erro ao atualizar categoria: {e}")
                return False
            finally:
                if cursor:
                    cursor.close()  # Garante que o cursor seja fechado

    def is_categoria_by_nome(self):
        conexao = Banco.getConexao()
        if conexao:
            try:
                cursor = conexao.cursor()
                sql = "SELECT COUNT(*) FROM categorias WHERE nome_categoria = %s"
                cursor.execute(sql, (self._nome_categoria,))
                result = cursor.fetchone()
                return result[0] > 0
            except Error as e:
                print(f"Erro ao verificar categoria por nome: {e}")
                return False
            finally:
                cursor.close()

    def is_categoria_by_id(self, id_categoria):
        conexao = Banco.getConexao()
        if conexao:
            try:
                cursor = conexao.cursor()
                sql = "SELECT COUNT(*) FROM categorias WHERE id_categoria = %s"
                cursor.execute(sql, (id_categoria,))
                result = cursor.fetchone()
                return result[0] > 0 
            except Error as e:
                print(f"Erro ao verificar categoria por ID: {e}")
                return False
            finally:
                cursor.close()

    def read_all(self):
        conexao = Banco.getConexao()
        if conexao:
            try:
                cursor = conexao.cursor(dictionary=True)
                sql = "SELECT * FROM categorias ORDER BY nome_categoria"
                cursor.execute(sql)
                return cursor.fetchall()
            except Error as e:
                print(f"Erro ao obter categorias: {e}")
                return []
            finally:
                cursor.close()

    def read_by_id(self):
        conexao = Banco.getConexao()
        if conexao:
            try:
                cursor = conexao.cursor(dictionary=True)
                sql = "SELECT * FROM categorias WHERE id_categoria = %s"
                cursor.execute(sql, (self._id_categoria,))
                result = cursor.fetchone()
                return result 
            except Error as e:
                print(f"Erro ao obter categoria por ID: {e}")
                return None
            finally:
                cursor.close()

    @property
    def id_categoria(self):
        return self._id_categoria

    @id_categoria.setter
    def id_categoria(self, value):
        self._id_categoria = value

    @property
    def nome_categoria(self):
        return self._nome_categoria

    @nome_categoria.setter
    def nome_categoria(self, value):
        self._nome_categoria = value

    @property
    def desconto_categoria(self):
        return self._desconto_categoria

    @desconto_categoria.setter
    def desconto_categoria(self, value):
        self._desconto_categoria = value

    @property
    def id_setor(self):
        return self._id_setor

    @id_setor.setter
    def id_setor(self, value):
        self._id_setor = value

File: Node/model/Categoria.js
// Importa o módulo Banco para realizar conexões com o banco de dados.
const Banco = require('./Banco');

// Define a classe Categoria para representar a entidade Categoria.
class Categoria {
    // Construtor da classe Categoria que inicializa as propriedades.
    constructor() {
        this._id_categoria = null;  // ID da categoria, inicialmente nulo.
        this._nome_categoria = null;  // Nome do categoria, inicialmente uma string vazia.
        this._desconto_categoria = null;
        this._id_setor = null;
    }

    // Método assíncrono para criar um novo categoria no banco de dados.
    async create() {
        const conexao = Banco.getConexao();  // Obtém a conexão com o banco de dados.
        const SQL = 'INSERT INTO categorias (nome_categoria, desconto_categoria, id_setor) VALUES (?, ?, ?);';  // Query SQL para inserir o nome do categoria.

        try {
            const [result] = await conexao.promise().execute(SQL, [this._nome_categoria, this._desconto_categoria, this._id_setor]);  // Executa a query.
            this._id_categoria = result.insertId;  // Armazena o ID gerado pelo banco de dados.
            return result.affectedRows > 0;  // Retorna true se a inserção afetou alguma linha.
        } catch (error) {
            console.error('Erro ao criar o categoria:', error);  // Exibe erro no console se houver falha.
            return false;  // Retorna false caso ocorra um erro.
        }
    }

    // Método assíncrono para excluir um categoria do banco de dados.
    async delete() {
        const conexao = Banco.getConexao();  // Obtém a conexão com o banco de dados.
        const SQL = 'DELETE FROM categorias WHERE id_categoria = ?;';  // Query SQL para deletar um categoria pelo ID.

        try {
            const [result] = await conexao.promise().execute(SQL, [this._id_categoria]);  // Executa a query de exclusão.
            return result.affectedRows > 0;  // Retorna true se alguma linha foi afetada (categoria deletado).
        } catch (error) {
            console.error('Erro ao excluir o categoria:', error);  // Exibe erro no console se houver falha.
            return false;  // Retorna false caso ocorra um erro.
        }
    }

    // Método assíncrono para atualizar os dados de um categoria no banco de dados.
    async update() {
        const conexao = Banco.getConexao();  // Obtém a conexão com o banco de dados.
        const SQL = 'UPDATE categorias SET nome_categoria = ?, desconto_categoria = ?, id_setor = ? WHERE id_categoria = ?;';  // Query SQL para atualizar o nome de um categoria.

        try {
            const [result] = await conexao.promise().execute(SQL, [this._nome_categoria, this._desconto_categoria, this._id_setor, this._id_categoria]);  // Executa a query de atualização.
            return result.affectedRows > 0;  // Retorna true se a atualização afetou alguma linha.
        } catch (error) {
            console.error('Erro ao atualizar o categoria:', error);  // Exibe erro no console se houver falha.
            return false;  // Retorna false caso ocorra um erro.
        }
    }

    // Método assíncrono para verificar se um categoria já existe no banco de dados.
    async isCategoriaByNomeCategoria() {
        const conexao = Banco.getConexao();  // Obtém a conexão com o banco de dados.
        const SQL = 'SELECT COUNT(*) AS qtd FROM categorias WHERE nome_categoria = ?;';  // Query SQL para contar categorias com o mesmo nome.

        try {
            const [rows] = await conexao.promise().execute(SQL, [this._nome_categoria]);  // Executa a query.
            return rows[0].qtd > 0;  // Retorna true se houver algum categoria com o mesmo nome.
        } catch (error) {
            console.error('Erro ao verificar o categoria:', error);  // Exibe erro no console se houver falha.
            return false;  // Retorna false caso ocorra um erro.
        }
    }

    async isCategoriaById(id_categoria) {
        const conexao = Banco.getConexao();  // Obtém a conexão com o banco de dados.
        const SQL = 'SELECT COUNT(*) AS qtd FROM categorias WHERE id_categoria = ?;';  // Query SQL para contar categorias com o mesmo nome.
        try {
            const [rows] = await conexao.promise().execute(SQL, [id_categoria]);  // Executa a query.
            return rows[0].qtd > 0;  // Retorna true se houver algum categoria com o mesmo nome.
        } catch (error) {
            console.error('Erro ao verificar o categoria:', error);  // Exibe erro no console se houver falha.
            return false;  // Retorna false caso ocorra um erro.
        }
    }

    // Método assíncrono para ler todos os categorias do banco de dados.
    async readAll() {
        const conexao = Banco.getConexao();  // Obtém a conexão com o banco de dados.
        const SQL = 'SELECT * FROM categorias ORDER BY nome_categoria;';  // Query SQL para selecionar todos os categorias ordenados pelo nome.

        try {
            const [rows] = await conexao.promise().execute(SQL);  // Executa a query de seleção.
            return rows;  // Retorna a lista de categorias.
        } catch (error) {
            console.error('Erro ao ler categorias:', error);  // Exibe erro no console se houver falha.
            return [];  // Retorna uma lista vazia caso ocorra um erro.
        }
    }

    // Método assíncrono para ler um categoria pelo seu ID.
    async readByID() {
        const conexao = Banco.getConexao();  // Obtém a conexão com o banco de dados.
        const SQL = 'SELECT * FROM categorias WHERE id_categoria = ?;';  // Query SQL para selecionar um categoria pelo ID.

        try {
            const [rows] = await conexao.promise().execute(SQL, [this._id_categoria]);  // Executa a query de seleção.
            return rows;  // Retorna o categoria correspondente ao ID.
        } catch (error) {
            console.error('Erro ao ler categoria pelo ID:', error);  // Exibe erro no console se houver falha.
            return null;  // Retorna null caso ocorra um erro.
        }
    }

    // Getter para obter o valor de id_categoria.
    get id_categoria() {
        return this._id_categoria;
    }

    // Setter para definir o valor de id_categoria.
    set id_categoria(id_categoria) {
        this._id_categoria = id_categoria;
        return this;  // Retorna a instância atual para permitir encadeamento de chamadas.
    }

    // Getter para obter o valor de nome_categoria.
    get nome_categoria() {
        return this._nome_categoria;
    }

    // Setter para definir o valor de nome_categoria.
    set nome_categoria(nome_categoria) {
        this._nome_categoria = nome_categoria;
        return this;  // Retorna a instância atual para permitir encadeamento de chamadas.
    }

    // Getter para obter o valor de id_setor.
    get id_setor() {
        return this._id_setor;
    }

    // Setter para definir o valor de id_setor.
    set id_setor(id_setor) {
        this._id_setor = id_setor;
        return this;  // Retorna a instância atual para permitir encadeamento de chamadas.
    }

    // Getter para obter o valor de desconto_categoria.
    get desconto_categoria() {
        return this._desconto_categoria;
    }

    // Setter para definir o valor de desconto_categoria.
    set desconto_categoria(desconto_categoria) {
        this._desconto_categoria = desconto_categoria;
        return this;  // Retorna a instância atual para permitir encadeamento de chamadas.
    }
}

// Exporta a classe Categoria para que possa ser utilizada em outros módulos.
module.exports = Categoria;
